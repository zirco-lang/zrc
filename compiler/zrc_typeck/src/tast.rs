//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.
//!
//! Note that we only attach source locations to code that would exist after code generation,
//! meaning inlined types do not get a span.

pub mod expr;
pub mod stmt;
pub mod ty;
