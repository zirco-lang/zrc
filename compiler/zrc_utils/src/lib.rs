#![doc=include_str!("../README.md")]
#![allow(unknown_lints)] // in case you use non-nightly clippy
#![warn(
    clippy::cargo,
    clippy::nursery,
    clippy::pedantic,
    clippy::missing_docs_in_private_items,
    missing_docs,
    clippy::absolute_paths,
    clippy::as_conversions,
    clippy::dbg_macro,
    clippy::decimal_literal_representation,
    clippy::deref_by_slicing,
    clippy::disallowed_script_idents,
    clippy::else_if_without_else,
    clippy::empty_structs_with_brackets,
    clippy::format_push_string,
    clippy::if_then_some_else_none,
    clippy::let_underscore_must_use,
    clippy::min_ident_chars,
    clippy::mixed_read_write_in_expression,
    clippy::multiple_inherent_impl,
    clippy::multiple_unsafe_ops_per_block,
    clippy::non_ascii_literal,
    clippy::redundant_type_annotations,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::same_name_method,
    clippy::semicolon_inside_block,
    clippy::unseparated_literal_suffix,
    clippy::string_to_string,
    clippy::todo,
    clippy::undocumented_unsafe_blocks,
    clippy::unimplemented,
    clippy::unneeded_field_pattern,
    clippy::wildcard_enum_match_arm,

    // These should be enabled in any non-user-facing code, like the parser, but not in the
    // frontend.
    clippy::print_stderr,
    clippy::print_stdout
)]
#![allow(clippy::multiple_crate_versions, clippy::cargo_common_metadata)]

pub mod span;
