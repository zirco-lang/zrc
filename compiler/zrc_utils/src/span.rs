//! Associations to spans within the input source
//!
//! This module declares two very helpful types, [`Span`] and [`Spanned<T>`],
//! which both help the compiler associate things like tokens with locations
//! within the source file. There is also a trait [`Spannable`] which allows
//! some easier chained method calls.
//!
//! Read the documentation for the types [`Span`] and [`Spanned<T>`], and the
//! trait [`Spannable`] to learn more.

use std::{fmt::Display, ops::RangeInclusive};

/// Represents the start and end of some segment of a string
///
/// A span represents the start and end of some span within a string. It can be
/// sliced into a str, and various other operations. You can also create a
/// [`Spanned<T>`] from a span using the [`Span::containing`] method.
///
/// These are often found within the [`Spanned<T>`] type. They can be obtained
/// in a few ways:
/// - Direct construction ([`Span::from_positions`])
/// - Methods on another Span ([`Span::intersect`])
/// - Stripping the value from a [`Spanned<T>`] ([`Spanned::span`])
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Span(usize, usize);
impl Span {
    /// Create a new [`Span`] given a start and end location.
    ///
    /// # Panics
    /// Panics if `start > end`.
    #[must_use]
    pub fn from_positions(start: usize, end: usize) -> Self {
        assert!(end >= start, "span must have positive length");
        Self(start, end)
    }

    /// Obtains the starting position of this [`Span`] as a `usize`
    #[must_use]
    pub const fn start(&self) -> usize {
        self.0
    }

    /// Obtains the ending position of this [`Span`] as a `usize`
    #[must_use]
    pub const fn end(&self) -> usize {
        self.1
    }

    /// Convert this [`Span`] into a [`RangeInclusive`], good for slicing into
    /// your input
    #[must_use]
    pub const fn range(&self) -> RangeInclusive<usize> {
        self.start()..=self.end()
    }

    /// Creates a [`Spanned<T>`] instance using this [`Span`] and a passed value
    #[must_use]
    pub const fn containing<T>(self, value: T) -> Spanned<T> {
        Spanned::from_span_and_value(self, value)
    }

    /// Creates a new [`Span`] containing the intersection of two passed spans
    ///
    /// If this returns [`None`], no intersection exists (they are disjoint).
    #[must_use]
    pub fn intersect(span_a: Self, span_b: Self) -> Option<Self> {
        if span_a.start() > span_b.end() || span_b.start() > span_a.end() {
            None
        } else {
            Some(Self::from_positions(
                std::cmp::max(span_a.start(), span_b.start()),
                std::cmp::min(span_a.end(), span_b.end()),
            ))
        }
    }
}
impl Display for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}-{}", self.start(), self.end())
    }
}

/// Represents something (`T`) contained within a [`Span`].
///
/// These are often found in many places throughout the compiler, such as
/// attached to AST nodes or in diagnostics.
///
/// An instance of [`Spanned`] can be obtained in a few ways:
/// - Direct construction from a [`Span`] and a value
///   ([`Spanned::from_span_and_value`])
/// - By attaching a value to a [`Span`] ([`Span::containing`])
/// - By attaching a [`Span`] to a value (with the [`Spannable`] trait's
///   [`Spannable::in_span`] method)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Spanned<T>(Span, T);
impl<T> Spanned<T> {
    /// Create a new [`Spanned<T>`] instance from a [`Span`] and some value
    #[must_use]
    pub const fn from_span_and_value(span: Span, value: T) -> Self {
        Self(span, value)
    }

    /// Obtains the [`Span`] associated with this [`Spanned<T>`] instance
    #[must_use]
    pub const fn span(&self) -> Span {
        self.0
    }

    /// Obtains a reference to the value this [`Spanned<T>`] instance wraps
    pub const fn value(&self) -> &T {
        &self.1
    }

    /// Applies a function to the contained value, returning a new
    /// [`Spanned<T>`] instance with the same associated [`Span`]
    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
        self.span().containing(f(self.into_value()))
    }

    /// "Strips" the [`Spanned<T>`] of its [`Span`], returning the inner value
    ///
    /// This differs from [`Spanned::value`] because it consumes the
    /// [`Spanned<T>`] instance and drops the [`Span`].
    #[allow(clippy::missing_const_for_fn)]
    pub fn into_value(self) -> T {
        self.1
    }

    /// Obtains the starting position of the contained [`Span`] as a `usize`
    #[must_use]
    pub const fn start(&self) -> usize {
        self.span().start()
    }

    /// Obtains the ending position of the contained [`Span`] as a `usize`
    #[must_use]
    pub const fn end(&self) -> usize {
        self.span().end()
    }

    /// Converts a [`&Spanned<T>`][Spanned] to a [`Spanned<&T>`].
    pub const fn as_ref(&self) -> Spanned<&T> {
        self.span().containing(&self.1)
    }
}
impl<T> Spanned<Option<T>> {
    /// Converts a [`Spanned<Option<T>>`] to a [`Option<Spanned<T>>`].
    /// Note: This is not reversible because if you wanted to create a
    /// [`Spanned`] [`Some`] from [`None`], what span would you use?
    pub fn transpose(self) -> Option<Spanned<T>> {
        let span = self.span();
        self.into_value().map(|x| x.in_span(span))
    }
}
impl<T, E> Spanned<Result<T, E>> {
    /// Converts a [`Spanned<Result<T, E>>`] to a [`Result<Spanned<T>,
    /// Spanned<E>>`]. Note: This is not reversible. See the note on
    /// [`Spanned<Option<T>>::transpose`].
    #[allow(clippy::missing_errors_doc)] // just propagates input error
    pub fn transpose(self) -> Result<Spanned<T>, Spanned<E>> {
        let span = self.span();
        self.into_value()
            .map(|x| x.in_span(span))
            .map_err(|x| x.in_span(span))
    }
}

/// A trait automatically implemented on all types that allows you to attach a
/// [`Span`] to something, creating a [`Spanned<T>`] instance.
pub trait Spannable
where
    Self: Sized,
{
    /// Attach a [`Span`] to this value, creating a [`Spanned<T>`] instance
    ///
    /// This method can be used to attach a [`Span`] to any arbitrary value. It
    /// is a cleaner syntax for the [`Spanned::from_span_and_value`] or
    /// [`Span::containing`] functions.
    fn in_span(self, span: Span) -> Spanned<Self>;
}

// Automatically implement Spannable for all types
impl<T: Sized> Spannable for T {
    fn in_span(self, span: Span) -> Spanned<Self> {
        Spanned::from_span_and_value(span, self)
    }
}

/// Create a [`Spanned<T>`] instance from two locations and a value.
/// Simply just expands to a [`Spanned::from_span_and_value`] and
/// [`Span::from_positions`] calls.
///
/// # Panics
/// Panics if `start > end`.
#[macro_export]
macro_rules! spanned {
    ($start:expr, $value:expr, $end:expr) => {
        $crate::span::Spanned::from_span_and_value(
            $crate::span::Span::from_positions($start, $end),
            $value,
        )
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn spanned_macro_creates_spanned_item() {
        assert_eq!(
            spanned!(0, (), 3),
            Spanned::from_span_and_value(Span::from_positions(0, 3), ())
        );
    }

    mod span {
        use super::*;

        #[test]
        fn span_from_positions_works_as_expected() {
            let span = Span::from_positions(2, 7);
            assert_eq!(span, Span(2, 7));
            assert_eq!(span.start(), 2);
            assert_eq!(span.end(), 7);
            assert_eq!(span.range(), 2..=7);
            assert_eq!(span.to_string(), "2-7".to_string());
        }

        #[test]
        #[allow(clippy::let_underscore_must_use)]
        #[should_panic(expected = "span must have positive length")]
        fn span_from_invalid_positions_panics() {
            let _ = Span::from_positions(5, 0);
        }

        #[test]
        fn span_intersection_with_overlap_returns_intersection() {
            assert_eq!(
                Span::intersect(Span::from_positions(0, 6), Span::from_positions(4, 10)),
                Some(Span::from_positions(4, 6))
            );
        }

        #[test]
        fn span_intersections_with_disjoint_spans_returns_none() {
            assert_eq!(
                Span::intersect(Span::from_positions(0, 5), Span::from_positions(7, 10)),
                None
            );
        }

        #[test]
        fn span_containing_returns_spanned() {
            assert_eq!(
                Span::from_positions(0, 3).containing(()),
                spanned!(0, (), 3)
            );
        }
    }

    mod spanned {
        use super::*;

        #[test]
        fn basic_methods_work_as_expected() {
            let span = Span::from_positions(3, 6);
            let spanned = Spanned::from_span_and_value(span, 0);

            assert_eq!(spanned.span(), span);
            assert_eq!(spanned.start(), 3);
            assert_eq!(spanned.end(), 6);
            assert_eq!(spanned.value(), &0);
            assert_eq!(spanned.into_value(), 0);
        }

        #[test]
        fn map_works_as_expected() {
            let spanned = spanned!(3, 0, 6);

            assert_eq!(spanned.map(|n| n == 0), spanned!(3, true, 6));
        }

        #[test]
        fn as_ref_works_as_expected() {
            let spanned = spanned!(3, 0, 6);

            assert_eq!(spanned.as_ref(), spanned!(3, &0, 6));
        }

        #[test]
        fn transpose_option_some_case() {
            let spanned = spanned!(3, Some(0), 6);

            assert_eq!(spanned.transpose(), Some(spanned!(3, 0, 6)));
        }

        #[test]
        fn transpose_option_none_case() {
            let spanned: Spanned<Option<()>> = spanned!(3, None, 6);

            assert_eq!(spanned.transpose(), None);
        }

        #[test]
        fn transpose_result_ok_case() {
            let spanned: Spanned<Result<i32, ()>> = spanned!(3, Ok(0), 6);

            assert_eq!(spanned.transpose(), Ok(spanned!(3, 0, 6)));
        }

        #[test]
        fn transpose_result_err_case() {
            let spanned: Spanned<Result<(), i32>> = spanned!(3, Err(0), 6);

            assert_eq!(spanned.transpose(), Err(spanned!(3, 0, 6)));
        }
    }

    #[test]
    fn spannable_in_span_creates_spanned() {
        assert_eq!(
            7.in_span(Span::from_positions(3, 6)),
            Spanned(Span(3, 6), 7),
        );
    }
}
