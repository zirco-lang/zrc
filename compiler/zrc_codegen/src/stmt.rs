//! Code generation for statements

use inkwell::{
    basic_block::BasicBlock,
    debug_info::{AsDIScope, DILexicalBlock},
};
use zrc_typeck::tast::{
    expr::{Place, PlaceKind, TypedExpr, TypedExprKind},
    stmt::{LetDeclaration, TypedStmt, TypedStmtKind},
    ty::Type,
};
use zrc_utils::span::{Span, Spannable, Spanned};

use crate::{
    ctx::{BlockCtx, FunctionCtx},
    expr::cg_expr,
    ty::llvm_basic_type,
    BasicBlockAnd, CgScope,
};

/// Consists of the [`BasicBlock`]s to `br` to when encountering certain
/// instructions. It is passed to [`cg_block`] to allow it to properly handle
/// break and continue.
#[derive(PartialEq, Eq, Debug, Clone)]
#[allow(clippy::redundant_pub_crate)]
pub(crate) struct LoopBreakaway<'ctx> {
    /// Points to the exit basic block.
    on_break: BasicBlock<'ctx>,
    /// For `for` loops, points to the latch. For `while` loops, points to the
    /// header.
    on_continue: BasicBlock<'ctx>,
}

/// Generates the `alloca`tion, `store` instruction, and adds a new identifier
/// to the [`CgScope`].
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(clippy::too_many_arguments)]
fn cg_let_declaration<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    mut bb: BasicBlock<'ctx>,
    scope: &'a mut CgScope<'input, 'ctx>,
    dbg_scope: DILexicalBlock<'ctx>,
    declarations: Vec<Spanned<LetDeclaration<'input>>>,
) -> BasicBlock<'ctx> {
    for spanned_let_declaration in declarations {
        let span = spanned_let_declaration.span();
        let let_declaration = spanned_let_declaration.into_value();

        let stmt_line_col = cg.line_lookup.lookup_from_index(span.start());
        let debug_location = cg.dbg_builder.create_debug_location(
            cg.ctx,
            stmt_line_col.line,
            stmt_line_col.col,
            dbg_scope.as_debug_info_scope(),
            None,
        );
        cg.builder.set_current_debug_location(debug_location);

        // we create our own builder here because we need to insert the alloca
        // at the beginning of the entry block, and that is easier than trying to
        // somehow save our position.

        let entry_block_builder = cg.ctx.create_builder();
        let first_bb = cg
            .fn_value
            .get_first_basic_block()
            .expect("function should have at least one basic block");

        #[allow(clippy::option_if_let_else)]
        match first_bb.get_first_instruction() {
            Some(first_instruction) => {
                entry_block_builder.position_before(&first_instruction);
            }
            None => {
                entry_block_builder.position_at_end(first_bb);
            }
        }

        let (ty, dbg_ty) = llvm_basic_type(&cg, &let_declaration.ty);

        let ptr = entry_block_builder
            .build_alloca(ty, &format!("let_{}", let_declaration.name))
            .expect("alloca should generate successfully");

        scope.insert(let_declaration.name.value(), ptr);

        let decl = cg.dbg_builder.create_auto_variable(
            dbg_scope.as_debug_info_scope(),
            let_declaration.name.value(),
            cg.compilation_unit.get_file(),
            cg.line_lookup.lookup_from_index(span.start()).line,
            dbg_ty,
            true,
            0,
            0,
        );

        cg.dbg_builder
            .insert_declare_at_end(ptr, Some(decl), None, debug_location, first_bb);

        if let Some(value) = let_declaration.value {
            let expr_cg = BlockCtx::new(cg, scope, dbg_scope);

            bb = cg_expr(
                expr_cg,
                bb,
                TypedExpr {
                    inferred_type: let_declaration.ty.clone(),
                    kind: value.kind.span().containing(TypedExprKind::Assignment(
                        Box::new(Place {
                            inferred_type: let_declaration.ty,
                            kind: PlaceKind::Variable(let_declaration.name.value())
                                .in_span(let_declaration.name.span()),
                        }),
                        Box::new(value),
                    )),
                },
            )
            .bb;
        }
    }

    bb
}

/// Process a vector of [`TypedStmt`]s (a block) and handle each statement.
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(
    clippy::too_many_arguments,
    clippy::too_many_lines,
    clippy::needless_pass_by_value,
    clippy::redundant_pub_crate
)]
pub(crate) fn cg_block<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    bb: BasicBlock<'ctx>,
    parent_scope: &'a CgScope<'input, 'ctx>,
    parent_lexical_block: DILexicalBlock<'ctx>,
    block: Spanned<Vec<TypedStmt<'input>>>,
    breakaway: &Option<LoopBreakaway<'ctx>>,
) -> Option<BasicBlock<'ctx>> {
    let mut scope = parent_scope.clone();
    let block_span = block.span();
    let block_line_col = cg.line_lookup.lookup_from_index(block_span.start());
    let lexical_block = cg.dbg_builder.create_lexical_block(
        parent_lexical_block.as_debug_info_scope(),
        cg.compilation_unit.get_file(),
        block_line_col.line,
        block_line_col.col,
    );

    block
        .into_value()
        .into_iter()
        .try_fold(bb, |bb, stmt| -> Option<BasicBlock> {
            let stmt_span = stmt.0.span();
            let stmt_line_col = cg.line_lookup.lookup_from_index(stmt_span.start());
            let debug_location = cg.dbg_builder.create_debug_location(
                cg.ctx,
                stmt_line_col.line,
                stmt_line_col.col,
                lexical_block.as_debug_info_scope(),
                None,
            );
            cg.builder.set_current_debug_location(debug_location);

            match stmt.0.into_value() {
                TypedStmtKind::ExprStmt(expr) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    Some(cg_expr(expr_cg, bb, expr).bb)
                }

                TypedStmtKind::IfStmt(cond, then, then_else) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let then_else = then_else.unwrap_or_else(|| {
                        vec![].in_span(Span::from_positions(then.end(), then.end()))
                    });

                    let then_end = then.end();
                    let then_else_end = then_else.end();

                    let cond = cg_expr(expr_cg, bb, cond).into_value();

                    let then_bb = cg.ctx.append_basic_block(cg.fn_value, "then");
                    let then_else_bb = cg.ctx.append_basic_block(cg.fn_value, "then_else");

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), then_bb, then_else_bb)
                        .expect("conditional branch should generate successfully");

                    cg.builder.position_at_end(then_bb);
                    let maybe_then_bb =
                        cg_block(cg, then_bb, &scope, lexical_block, then, breakaway);

                    cg.builder.position_at_end(then_else_bb);
                    let maybe_then_else_bb = cg_block(
                        cg,
                        then_else_bb,
                        &scope,
                        lexical_block,
                        then_else,
                        breakaway,
                    );

                    match (maybe_then_bb, maybe_then_else_bb) {
                        (None, None) => None,
                        (Some(single_bb), None) | (None, Some(single_bb)) => {
                            let end = cg.ctx.append_basic_block(cg.fn_value, "end");

                            let then_end_line_col = cg.line_lookup.lookup_from_index(then_end);
                            let terminating_debug_location = cg.dbg_builder.create_debug_location(
                                cg.ctx,
                                then_end_line_col.line,
                                then_end_line_col.col,
                                lexical_block.as_debug_info_scope(),
                                None,
                            );

                            cg.builder
                                .set_current_debug_location(terminating_debug_location);

                            cg.builder.position_at_end(single_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            cg.builder.position_at_end(end);
                            Some(end)
                        }
                        (Some(then_bb), Some(then_else_bb)) => {
                            let end = cg.ctx.append_basic_block(cg.fn_value, "end");

                            let then_end_line_col = cg.line_lookup.lookup_from_index(then_end);
                            let then_terminating_debug_location =
                                cg.dbg_builder.create_debug_location(
                                    cg.ctx,
                                    then_end_line_col.line,
                                    then_end_line_col.col,
                                    lexical_block.as_debug_info_scope(),
                                    None,
                                );
                            cg.builder
                                .set_current_debug_location(then_terminating_debug_location);
                            cg.builder.position_at_end(then_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            let then_else_end_line_col =
                                cg.line_lookup.lookup_from_index(then_else_end);
                            let then_else_terminating_debug_location =
                                cg.dbg_builder.create_debug_location(
                                    cg.ctx,
                                    then_else_end_line_col.line,
                                    then_else_end_line_col.col,
                                    lexical_block.as_debug_info_scope(),
                                    None,
                                );
                            cg.builder
                                .set_current_debug_location(then_else_terminating_debug_location);

                            cg.builder.position_at_end(then_else_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            cg.builder.position_at_end(end);
                            Some(end)
                        }
                    }
                }

                TypedStmtKind::BlockStmt(block) => cg_block(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    block.in_span(stmt_span),
                    breakaway,
                ),

                TypedStmtKind::ReturnStmt(Some(expr)) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let expr = cg_expr(expr_cg, bb, expr).into_value();

                    cg.builder
                        .build_return(Some(&expr))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ReturnStmt(None) => {
                    let unit_type = llvm_basic_type(&cg, &Type::unit());

                    cg.builder
                        .build_return(Some(&unit_type.0.const_zero()))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ContinueStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `continue` is valid")
                                .on_continue,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::BreakStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `break` is valid")
                                .on_break,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::DeclarationList(declarations) => Some(cg_let_declaration(
                    cg,
                    bb,
                    &mut scope,
                    lexical_block,
                    declarations,
                )),

                TypedStmtKind::ForStmt {
                    init,
                    cond,
                    post,
                    body,
                } => {
                    // For loops generate a somewhat more complicated CFG, with a few parts.
                    // The preheader, where `init` runs. Breaks to the header.
                    // The header, where `cond` is checked and breaks to either the exit or the
                    // body. The body, where most of the body runs. Breaks to
                    // the latch. `break` transfers to the exit by force and `continue` transfers to
                    // the latch by force. The latch, where `post` runs and
                    // breaks back to the header The exit, which is the basic
                    // block we return.

                    // loops lie in an implicit subscope
                    let mut scope = scope.clone();

                    // The block we are currently in will become the preheader. Generate the `init`
                    // code if there is any.
                    if let Some(init) = init {
                        cg_let_declaration(cg, bb, &mut scope, lexical_block, *init);
                    }

                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let header = cg.ctx.append_basic_block(cg.fn_value, "header");
                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
                    let latch = cg.ctx.append_basic_block(cg.fn_value, "latch");
                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    // Branch to the header from the preheader.
                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    // Generate the header.
                    cg.builder.position_at_end(header);
                    let header = cond.map_or_else(
                        || {
                            // If there is no condition, we always branch to the body.
                            cg.builder
                                .build_unconditional_branch(body_bb)
                                .expect("branch should generate successfully");

                            header
                        },
                        |cond| {
                            let mut header = header;

                            let cond = unpack!(header = cg_expr(expr_cg, header, cond));

                            cg.builder
                                .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                                .expect("branch should generate successfully");

                            header
                        },
                    );

                    // Generate the body.
                    cg.builder.position_at_end(body_bb);
                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: latch,
                        }),
                    );

                    // The body breaks to latch
                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(latch)
                            .expect("branch should generate successfully");
                    }

                    // Latch runs post and then breaks right back to the header.
                    cg.builder.position_at_end(latch);
                    if let Some(post) = post {
                        cg_expr(expr_cg, latch, post);
                    }

                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }

                TypedStmtKind::WhileStmt(cond, body) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    // While loops are similar to for loops but much simpler.
                    // The preheader simply just breaks to the header.
                    // The header checks the condition and breaks to the exit or the body.
                    // The body simply breaks to the header.
                    // The exit is the continued code

                    // `break` => exit
                    // `continue` => header

                    let mut header = cg.ctx.append_basic_block(cg.fn_value, "header");

                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");

                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(header);

                    let cond = unpack!(header = cg_expr(expr_cg, header, cond));

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(body_bb);

                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: header,
                        }),
                    );

                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(header)
                            .expect("branch should generate successfully");
                    }

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }

                TypedStmtKind::DoWhileStmt(body, cond) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    // `do..while` loops are slightly different from `while` loops.
                    // the preheader breaks directly to the *body* and forces it to run at
                    // least once. the body can then later break to the header which checks the
                    // condition and will loop or exit.

                    // `break` => exit
                    // `continue` => header

                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
                    let body_start = body_bb;

                    let header = cg.ctx.append_basic_block(cg.fn_value, "header");

                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    cg.builder
                        .build_unconditional_branch(body_bb)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(body_bb);

                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: header,
                        }),
                    );

                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(header)
                            .expect("branch should generate successfully");
                    }

                    cg.builder.position_at_end(header);

                    let BasicBlockAnd { value: cond, .. } = cg_expr(expr_cg, header, cond);

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), body_start, exit)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }
            }
        })
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn let_declarations_are_properly_generated() {
        cg_snapshot_test!(indoc! {"
            fn test() {
                // TEST: should allocate twice and assign to one.
                let a: i32;
                let b: i32 = 7;
            }
        "});
    }

    mod cg_block {
        use super::*;

        #[test]
        fn function_parameters_are_properly_generated() {
            cg_snapshot_test!(indoc! {"
                fn id(x: i32) -> i32 {
                    return x;
                }
            "});
        }

        mod conditionals {
            use super::*;

            #[test]
            fn if_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code paths join at the end
                        if (get_bool()) nop();

                        // TEST: code generation properly continues in the last block
                        nop();
                        return;
                    }
                "});
            }

            #[test]
            fn if_else_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code baths call nop().
                        if (get_bool()) nop();
                        else {
                            nop();
                            // TEST: this path diverges
                            return;
                        }

                        // TEST: code generation properly continues from the if_true block
                        nop();
                    }
                "});
            }

            #[test]
            fn if_else_statements_where_both_blocks_terminate_do_not_continue_generating() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: properly produces a conditional break over the call result
                        // and both code paths return (diverge). there should be no %end bb.
                        if (get_bool()) return;
                        else return;
                    }
                "});
            }
        }

        mod loops {
            use super::*;

            #[test]
            fn while_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        while (get_bool()) {
                            // TEST: break jumps to the `end` block
                            if (get_bool()) break;
                            else {
                                // TEST: continue jumps to the header block
                                if (get_bool()) continue;
                                // TEST: otherwise, we proceed
                                else {}
                            }

                            // TEST: the loop jumps back to the header block
                        }

                        // TEST: ...and code generation properly continues.
                        return;
                    }
                "});
            }

            #[test]
            fn for_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_int() -> i32;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        for (let i = 0; i < get_int(); i += 1) {
                            // TEST: break jumps to the `end` block
                            if (i > get_int()) break;
                            else {
                                // TEST: continue jumps to the latch block
                                if (i < get_int()) continue;
                                else {}
                            }

                            // TEST: the loop jumps to the latch block which jumps back to the
                            // header
                        }
                    }
                "});
            }

            #[test]
            fn do_while_loops_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper `do..while` loop structure is created
                        do {
                            get_bool(); // for fake side effects
                        } while (get_bool());
                    }
                "});
            }
        }
    }
}
