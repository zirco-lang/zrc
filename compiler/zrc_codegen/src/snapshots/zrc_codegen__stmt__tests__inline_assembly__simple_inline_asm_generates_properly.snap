---
source: compiler/zrc_codegen/src/stmt.rs
description: "fn test() {\n    // TEST: simple inline assembly with no operands\n    asm(\"nop\");\n}\n"
expression: resulting_ir
snapshot_kind: text
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define {} @test() !dbg !3 {
entry:
  call void asm sideeffect "nop", ""(), !dbg !8
  ret {} zeroinitializer, !dbg !11
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1, retainedNodes: !7)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct {}", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !DILocation(line: 3, column: 5, scope: !9)
!9 = distinct !DILexicalBlock(scope: !10, file: !2, line: 1, column: 11)
!10 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 11)
!11 = !DILocation(line: 4, column: 1, scope: !9)
